[BREAKING] Strict conversion

hrkrshnn opened the pull request:
Closes https://github.com/ethereum/solidity/issues/9939

Decision on:
- [x] Conversions between Contract type and integer type. Should we only allow `C(uint160)`?
- [x] Conversion between Enums. If an enum `E` is represented by type `uint8`, should `uint16(E)` etc be disallowed?

Disallowed conversions from `integer -> contract` 

The behaviour of enums is unchanged.

hrkrshnn commented on libsolidity/ast/Types.h:
```diff
@@ -398,10 +398,36 @@ class Type
+		return true;
+	}
+
+	bool mutable m_lossfullyConverted = false;
```
Not sure if the choice of mutable is right here. Looking for suggestions.

cameel replied:
I find it unintuitive that simply calling `lossfullyConvertible()` (which is a `const` function so it looks harmless) can mark the type as converted, even if no conversion actually took place.

Also, what if two values refer to the same type instance? I'd expect something like this on a value, not on its type.

Just a quick idea: maybe this flag would work better as an annotation on expressions in the AST?

hrkrshnn replied:
> Just a quick idea: maybe this flag would work better as an annotation on expressions in the AST?

Do you think this would work in the following case:

```solidity
		uint x = uint(int(-1));
		uint8 y = uint8(x);
```

Currently it would produce the following error:
```
Error: Explicit type conversion not allowed from "uint256" to "uint8". The type "uint256" already underwent a potentially lossfull conversion.
 --> /tmp/example.sol:4:13:
  |
4 | 		uint8 y = uint8(x);
  | 		          ^^^^^^^^

```

hrkrshnn replied:
Let me check if annotating expressions would work in this case.

cameel replied:
But why would you want an error here? `uint8(uint(int(-1)))` would work so it should work if you split it too. There are two separate explicit conversions here.

cameel commented on libsolidity/ast/Types.h:
```diff
@@ -398,10 +398,36 @@ class Type
+ * Base class for Types that can be explicitly converted from another type, which can potentially
+ * lead to a loss in precision.
+ */
+class LossfullyConvertible
```
Just a note about naming: I think it's "lossy" rather than "lossful" in this context so I'd rename it to:
```suggestion
class LossilyConvertible
```

hrkrshnn committed:
Stricter explicit conversion between types.

A type can be converted to another if the conversion requires at most one of the following: sign,
width, kind (int, address, bytesXX, etc.) For example, the conversion `uint16(int8)` is now disallowed.

hrkrshnn committed:
Tests/Docs after stricter explicit conversion.

chriseth commented:
Concerning the first question: Depending on how often this happens, I would disallow any conversion between contract and uint - it has to go through address.

leonardoalt commented:
Regarding the second question: for consistency, I think it should be required to `uint16(uint8(e))`

leonardoalt commented on docs/080-breaking-changes.rst:
```diff
@@ -46,6 +46,25 @@ New Restrictions
      represent a value in the enum.
 
+* There are new restrictions on explicit conversions between two different types. The conversion is
+  only allowed when there is at most one change in sign, width or 'kind' (``int``, ``address``,
```
I think here you could just say type instead of kind.

leonardoalt commented on docs/080-breaking-changes.rst:
```diff
@@ -46,6 +46,25 @@ New Restrictions
+  ``bytesNN``, etc.) For example, the conversion ``uint16(int8)`` is disallowed since the conversion
+  changes width (8 bits to 16 bits) and sign (signed integer to unsigned integer.) To get the
+  previous behaviour, use an intermediate conversion. In the previous example, this would be
+  ``uint16(uint8(int8))``. The following are some examples of conversions that are disallowed by
```
```suggestion
  ``uint16(uint8(int8))``, or ``uint16(int16(int8))``. The following are some examples of conversions that are disallowed by
```

leonardoalt commented:
Are the same rules already valid for implicit conversion?

leonardoalt commented on libsolidity/ast/Types.cpp:
```diff
@@ -443,13 +443,22 @@ BoolResult AddressType::isImplicitlyConvertibleTo(Type const& _other) const
 BoolResult AddressType::isExplicitlyConvertibleTo(Type const& _convertTo) const
 {
-	if (_convertTo.category() == category())
+	if ((_convertTo.category() == category()) || isImplicitlyConvertibleTo(_convertTo))
```
The first is ok because it would only change width, right?

leonardoalt replied:
Ah no, wouldn't `uint16(int8)` also give true here?

leonardoalt replied:
Nevermind these comments, completely ignored that this is Address only

leonardoalt commented on libsolidity/ast/Types.cpp:
```diff
@@ -566,12 +575,40 @@ BoolResult IntegerType::isImplicitlyConvertibleTo(Type const& _convertTo) const
+	}
+	else if (auto fixedBytesType = dynamic_cast<FixedBytesType const*>(&_convertTo))
+	{
+		if (!isSigned() && (numBits() == fixedBytesType->numBytes() * 8))
```
Some of the tests such as this and the one above are repeated in some places. I'm wondering it it'd make sense to extract them somehow. Otherwise this looks good too.

hrkrshnn replied:
I thought about extracting into something like `canExplicitlyConvert(Type& _a, Type& _b)` and do it there. Maybe I'll do it after fixing the tests.

hrkrshnn replied:
I was thinking about converting each of these comparisons into individual functions, for example `canConvert(IntegerType const& _a, FixedBytesType const& _b)`. Since the rules for converting `a -> b` and `b -> a` are the same, (provided both conversions are feasible) both the cases can call the same function, except in one case, the arguments will have to be flipped. So these checks are not repeated. But I couldn't find a nice way to implement it. So I'll leave it like this.

leonardoalt replied:
Ok, I think that's fine too

chriseth replied:
Some way to make symmetry explicit would be nice, but I also don't see an easy way to do it.

leonardoalt commented on test/libsolidity/syntaxTests/types/strict_explicit.sol:
```diff
@@ -0,0 +1,42 @@
+    function f() public pure {
+
+        uint16 a = uint16(uint8(int8(-1)));
+		a;
```
Indentation (and below)

hrkrshnn committed:
Test updates after disallowing Integer -> Contract conversion

leonardoalt commented on docs/080-breaking-changes.rst:
```diff
@@ -46,6 +46,26 @@ New Restrictions
+  that are disallowed by this rule. Note that, given types ``T`` and ``S``, the notation ``T(S)``
+  refers to the explicit conversion ``T(x)``, where ``x`` is any arbitrary variable of type ``S``.
+
+  - ``address(uint)`` and ``uint(address)``: converting both 'kind' and width. Replace this by
```
I still find this `'kind'` weird, especially because it's always single quoted

hrkrshnn replied:
A problem with using type would be: distinguishing the difference between `uint16` v/s `int8` and 'uint16` v/s `address`. Each pair has different type, but the latter one has different kind. 

But what do you think is a good replacement?

chriseth replied:
type category?

leonardoalt replied:
Agree with type category

hrkrshnn replied:
Replaced 'kind' by type-category.

leonardoalt commented on libsolidity/ast/Types.cpp:
```diff
@@ -566,12 +575,30 @@ BoolResult IntegerType::isImplicitlyConvertibleTo(Type const& _convertTo) const
+	else if (dynamic_cast<EnumType const*>(&_convertTo))
+		return true;
+	else if (_convertTo.category() == Category::FixedPoint)
+		return true;
```
Can we do something more "precise" for FixedPoint?

chriseth replied:
We should at least check signedness.

hrkrshnn replied:
I added a check for signedness.

leonardoalt commented on libsolidity/ast/Types.cpp:
```diff
@@ -566,12 +575,30 @@ BoolResult IntegerType::isImplicitlyConvertibleTo(Type const& _convertTo) const
+			return true;
+	}
+	else if (dynamic_cast<EnumType const*>(&_convertTo))
+		return true;
```
So no checking the actual size of enums? Not sure what was decided

hrkrshnn replied:
We are not. chris mentioned in the chat that this is undecided. What do you think we should do?

leonardoalt replied:
Personally I think they should have the same rules using the uint type the enum gets converted to (smallest uint that fits). However, apparently we're going to make them uint8 always, so it might make sense to already assume enum is uint8.

hrkrshnn replied:
I agree with this. However, @chriseth and @cameel were not very keen about this change.

chriseth replied:
At least we should not allow signed integer to enum conversion. Having to write `MyEnum(uint8(x))` is a bit cumbersome and I don't think it provides a benefit. Do at least literals work?

hrkrshnn replied:
Literals would work if the literal fits in the enum.

leonardoalt commented on test/libsolidity/syntaxTests/types/strict_explicit.sol:
```diff
@@ -0,0 +1,42 @@
+    function f() public pure {
+
+        uint16 a = uint16(uint8(int8(-1)));
+        a;
```
Why these?

hrkrshnn replied:
I wanted to avoid the unused local variable warning.

leonardoalt replied:
ah ok

chriseth commented on Changelog.md:
```diff
@@ -11,6 +11,7 @@ Breaking Changes:
  * Code Generator: Use ``revert`` with error signature ``Panic(uint256)`` and error codes instead of invalid opcode on failing assertions.
  * Type System: Explicit conversions from literals to integer type is as strict as implicit conversions.
  * Type System: Explicit conversions from literals to enums are only allowed if the value fits in the enum.
+ * Type System: Explicit conversions between two types are disallowed if it changes more than one of sign, width or kind.
```
```suggestion
 * Type System: Explicit conversions between two types are disallowed if it changes more than one of sign, width or kind at the same time.
```

chriseth commented on docs/080-breaking-changes.rst:
```diff
@@ -46,6 +46,26 @@ New Restrictions
   3. Explicit conversions between literals and enums are only allowed if the literal can
      represent a value in the enum.
 
+* There are new restrictions on explicit conversions between two different types. The conversion is
```
```suggestion
* There are new restrictions on explicit type conversions. The conversion is
```

chriseth commented on docs/080-breaking-changes.rst:
```diff
@@ -46,6 +46,26 @@ New Restrictions
+* There are new restrictions on explicit conversions between two different types. The conversion is
+  only allowed when there is at most one change in sign, width or type-category (``int``,
+  ``address``, ``bytesNN``, etc.) For example, the conversion ``uint16(int8)`` is disallowed since
+  the conversion changes width (8 bits to 16 bits) and sign (signed integer to unsigned integer.) To
```
```suggestion
  the conversion changes width (8 bits to 16 bits) and sign (signed integer to unsigned integer). To
```

chriseth commented on docs/080-breaking-changes.rst:
```diff
@@ -46,6 +46,26 @@ New Restrictions
+  ``address``, ``bytesNN``, etc.) For example, the conversion ``uint16(int8)`` is disallowed since
+  the conversion changes width (8 bits to 16 bits) and sign (signed integer to unsigned integer.) To
+  get the previous behaviour, use an intermediate conversion. In the previous example, this would be
+  ``uint16(uint8(int8))`` or ``uint16(int16(int8))``. The following are some examples of conversions
```
Note that the two ways to convert will produce different results for e.g. `-1`

hrkrshnn replied:
Added this.

chriseth commented on docs/types/value-types.rst:
```diff
@@ -785,7 +785,7 @@ Another example that uses external function types::
 
     contract OracleUser {
-        Oracle constant private ORACLE_CONST = Oracle(0x1234567); // known contract
+        Oracle constant private ORACLE_CONST = Oracle(address(0x1234567)); // known contract
```
Can we use a proper address instead?

hrkrshnn replied:
Fixed

chriseth commented on docs/control-structures.rst:
```diff
@@ -260,15 +260,15 @@ which only need to be created if there is a dispute.
             // can be pre-computed. It is just there for illustration.
             // You actually only need ``new D{salt: salt}(arg)``.
-            address predictedAddress = address(uint(keccak256(abi.encodePacked(
+            address predictedAddress = address(uint160(uint(keccak256(abi.encodePacked(
```
This is kind of horrible...

chriseth replied:
But yeah, someone should write a helper function and that's it.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -443,13 +443,22 @@ BoolResult AddressType::isImplicitlyConvertibleTo(Type const& _other) const
 BoolResult AddressType::isExplicitlyConvertibleTo(Type const& _convertTo) const
 {
-	if (_convertTo.category() == category())
+	if ((_convertTo.category() == category()) || isImplicitlyConvertibleTo(_convertTo))
```
The first is needed because of payable conversion?

hrkrshnn replied:
Yes. The function `implicitlyConvertibleTo` doesn't allow conversions from non-payable to payable.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -566,12 +575,30 @@ BoolResult IntegerType::isImplicitlyConvertibleTo(Type const& _convertTo) const
 
 BoolResult IntegerType::isExplicitlyConvertibleTo(Type const& _convertTo) const
 {
-	return _convertTo.category() == category() ||
```
What about implicit conversion?

hrkrshnn replied:
It's redundant as it's already included in if-branch where `convertTo` is an `IntegerType`.

chriseth replied:
Shouldn't we just for symmetry always check for implicit convertibility first?

hrkrshnn replied:
I added the check for implicit conversion.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -566,12 +575,30 @@ BoolResult IntegerType::isImplicitlyConvertibleTo(Type const& _convertTo) const
-		(_convertTo.category() == Category::FixedBytes && numBits() == dynamic_cast<FixedBytesType const&>(_convertTo).numBytes() * 8) ||
-		_convertTo.category() == Category::FixedPoint;
+	if (auto integerType = dynamic_cast<IntegerType const*>(&_convertTo))
+	{
+		if (
+			(numBits() == integerType->numBits()) ||
+			(isSigned() == integerType->isSigned())
+		)
+			return true;
+	}
```
Would
```suggestion
		return
			(numBits() == integerType->numBits()) ||
			(isSigned() == integerType->isSigned());
```
be easier to understand?

chriseth commented on test/libsolidity/GasCosts.cpp:
```diff
@@ -162,10 +162,10 @@ BOOST_AUTO_TEST_CASE(single_callvaluecheck)
 				a = b;
 			}
 			function f1(address b) public pure returns (uint c) {
-				return uint(b) + 2;
```
This will cause the computation to be done in uint160 - not an issue in this test ,but it might cause bugs in contracts in general.

chriseth commented on test/libsolidity/semanticTests/storage/packed_storage_signed.sol:
```diff
@@ -13,14 +13,14 @@ contract C {
             c = a * int8(120) * int8(121);
         }
-        x1 = uint256(a);
+        x1 = uint256(uint8(a));
```
```suggestion
        x1 = uint256(uint256(a));
```

hrkrshnn replied:
That is not allowed since we can't convert from `int8` to `uint256`.

chriseth replied:
Sorry, I meant `int256` of course

chriseth commented on test/libsolidity/semanticTests/types/type_conversion_cleanup.sol:
```diff
@@ -1,7 +1,7 @@
 contract Test {
-    function test() public returns (uint ret) { return uint(address(uint128(type(uint200).max))); }
+    function test() public returns (uint ret) { return uint(uint160(address(uint160(type(uint200).max)))); }
```
```suggestion
    function test() public returns (uint ret) { return uint(uint160(address(uint160(uint128(type(uint200).max))))); }
```

chriseth commented:
I'm wondering whether we should reconsider this - also see https://github.com/ethereum/solidity/issues/10317

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -566,12 +572,18 @@ BoolResult IntegerType::isImplicitlyConvertibleTo(Type const& _convertTo) const
+	else if (dynamic_cast<EnumType const*>(&_convertTo))
+		return true;
+	else if (auto fixedPointType = dynamic_cast<FixedPointType const*>(&_convertTo))
+		return (isSigned() == fixedPointType->isSigned());
```
please at least check the number of bits.

hrkrshnn replied:
Added a check for `numBits()`

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -975,6 +987,7 @@ BoolResult RationalNumberType::isExplicitlyConvertibleTo(Type const& _convertTo)
 	{
 		if (isNegative() || isFractional() || integerType()->numBits() > 160)
 			return false;
+		return true;
 	}
```
```suggestion
return !isNegative() && !isFractional() && intigerType()->numBIts <= 160;
```

hrkrshnn replied:
Fixed

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -1446,10 +1459,16 @@ BoolResult FixedBytesType::isImplicitlyConvertibleTo(Type const& _convertTo) con
+	else if (_convertTo.category() == Category::Address && numBytes() == 20)
+		return true;
+	else if (_convertTo.category() == Category::FixedPoint)
+		return true;
```
please compare the sizes.

hrkrshnn replied:
Fixed

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -2674,7 +2693,11 @@ size_t EnumType::numberOfMembers() const
-	return _convertTo == *this || _convertTo.category() == Category::Integer;
+	if (_convertTo == *this)
+		return true;
+	else if (dynamic_cast<IntegerType const*>(&_convertTo))
```
not to signed

hrkrshnn replied:
Fixed

chriseth commented:
I would propose to disallow conversions between fixed point types as a breaking change so we have more freedom if we introduce them in a non-breaking change.

Also - are explicit conversions now fully symmetric? If yes, we could maybe extract that into isolated functions.

chriseth commented on docs/080-breaking-changes.rst:
```diff
@@ -49,6 +49,27 @@ New Restrictions
 
+* There are new restrictions on explicit type conversions. The conversion is only allowed when there
+  is at most one change in sign, width or type-category (``int``, ``address``, ``bytesNN``, etc.)
+  For example, the conversion ``uint16(int8)`` is disallowed since the conversion changes width (8
```
```suggestion
  For example, the conversion ``int8`` to ``uint16`` is disallowed since the conversion changes width (8
```

chriseth commented on docs/080-breaking-changes.rst:
```diff
@@ -49,6 +49,27 @@ New Restrictions
+* There are new restrictions on explicit type conversions. The conversion is only allowed when there
+  is at most one change in sign, width or type-category (``int``, ``address``, ``bytesNN``, etc.)
+  For example, the conversion ``uint16(int8)`` is disallowed since the conversion changes width (8
+  bits to 16 bits) and sign (signed integer to unsigned integer). To get the previous behaviour, use
```
```suggestion
  bits to 16 bits) and sign (signed integer to unsigned integer). In order to do the conversion,
  you have to go through an intermediate type.
```

chriseth commented on docs/080-breaking-changes.rst:
```diff
@@ -49,6 +49,27 @@ New Restrictions
+  is at most one change in sign, width or type-category (``int``, ``address``, ``bytesNN``, etc.)
+  For example, the conversion ``uint16(int8)`` is disallowed since the conversion changes width (8
+  bits to 16 bits) and sign (signed integer to unsigned integer). To get the previous behaviour, use
+  an intermediate conversion. In the previous example, this would be ``uint16(uint8(int8))`` or
```
```suggestion
  an intermediate conversion. In the previous example, this would be ``int8 -> uint8 -> uint16`` or
```

chriseth commented on docs/080-breaking-changes.rst:
```diff
@@ -49,6 +49,27 @@ New Restrictions
+  For example, the conversion ``uint16(int8)`` is disallowed since the conversion changes width (8
+  bits to 16 bits) and sign (signed integer to unsigned integer). To get the previous behaviour, use
+  an intermediate conversion. In the previous example, this would be ``uint16(uint8(int8))`` or
+  ``uint16(int16(int8))``. Note that the two ways to convert will produce different results e.g.,
```
```suggestion
  ``int8 -> int16 -> uint16``. Note that the two ways to convert will produce different results e.g.,
```

chriseth commented on docs/080-breaking-changes.rst:
```diff
@@ -49,6 +49,27 @@ New Restrictions
+  an intermediate conversion. In the previous example, this would be ``uint16(uint8(int8))`` or
+  ``uint16(int16(int8))``. Note that the two ways to convert will produce different results e.g.,
+  for ``-1``. The following are some examples of conversions that are disallowed by this rule. Note
+  that, given types ``T`` and ``S``, the notation ``T(S)`` refers to the explicit conversion
```
We can also use this notation above if you move the explanation up.

hrkrshnn replied:
Then, I'll just move the explanation up.

hrkrshnn commented on docs/080-breaking-changes.rst:
```diff
@@ -49,6 +49,28 @@ New Restrictions
+* There are new restrictions on explicit type conversions. The conversion is only allowed when there
+  is at most one change in sign, width or type-category (``int``, ``address``, ``bytesNN``, etc.)
+
+  Let us use the notation ``T(S)`` to denote the explicit conversion ``T(x)``, where, ``T`` and
```
I moved the text about documentation and did minor rewordings.

I'll update the conversions on fixed point types.

chriseth commented on docs/080-breaking-changes.rst:
```diff
@@ -49,6 +49,28 @@ New Restrictions
+
+  Let us use the notation ``T(S)`` to denote the explicit conversion ``T(x)``, where, ``T`` and
+  ``S`` are types, and ``x`` is any arbitrary variable of type ``S``. The conversion
+  ``uint16(int8)`` is now disallowed since the conversion changes width (8 bits to 16 bits) and sign
```
Can you make it clear that this is just an example? It could be read as the main change.

hrkrshnn replied:
Updated.
